core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -2010,10 +2012,12 @@ private[spark] object BlockManager {
 
       val filePath = file.path()
 
+      val deleter: DownloadFileDeleter = file.deleter()
+
       def cleanUp(): Unit = {
         logDebug(s"Clean up file $filePath")
 
-        if (!file.delete()) {
+        if (!deleter.delete()) {
```

Comment:
I'm probably missing something obvious but I'm not clear why the interface changes when things get deleted. It seems like it happens in the same place everywhere?

Reply:
The class extends `WeakReference` to track when `file: DownloadFile` gets garbage-collected, but at the same time it uses the reference in the `cleanUp` method which forces the compiler to create a field in `ReferenceWithCleanup` for it so it's never garbage-collected. The solution is to have a special deleter which can clean the underlying resources, but doesn't keep a reference to its `DownloadFile`.

Technically we could just replace `if (!file.delete()) {` with `if (!new File(filePath).delete()) {` and it would work, but the interface states that `DownloadFile::path` is intended for debug purposes only. I think the `path` method should be removed altogether or replaced with an alternative, like `toString()` or `description` so it doesn't leak implementation details.

I didn't want to create a huge PR, but I think in addition to removing `path` method we also should
- remove `referenceBuffer` (I don't see why we have it)
- move to scala's WeakReference and ReferenceQueue
(I was going to propose it in another PR)

Reply:
I get it. From the interface comments, I think it's OK to use `path` if that's a simpler solution, and just remove the comment. It's not going away. 

I agree with removing referenceBuffer. It's mutated but never queried.

Reply:
Sounds good to me.
I'll make the changes.

Reply:
Is that OK if I remove `DownloadFile::delete` method? It seems a bit weird to have the path exposed in the interface which is used for deletion and logging, and redundantly have another method to just delete the file.

Reply:
Done. Please let me know if I need to remove the second commit (the removal of `delete` method).

Reply:
FYI: I had to keep the `referenceBuffer` otherwise instances of `ReferenceWithCleanup` get garbage-collected too and cleanup never happens.